src/Service/AddressHelper.php

```diff
@@ -0,0 +1,135 @@
+            $this->entityManager->flush();
+        }
+
+        return !empty($addressData);
+    }
+
+    private function getAddress($entity, string $property): Address
+    {
+        $address = $this->propertyAccessor->getValue($entity, $property);
+        if (!($address instanceof Address)) {
+            throw $this->createException(sprintf('Property %s.%s must be an instance of %s; is %s', get_class($entity), $property, Address::class, get_class($address)));
+        }
+
+        return $address;
+    }
+
+    private function createException(string $message, $code = 0, Throwable $previous = null)
+    {
+        $this->logger->error($message, ['previous' => $previous]);
+
+        return new AddressException($message, $code, $previous);
+    }
+
+    /**
+     * Fetch address data using https://dawadocs.dataforsyningen.dk/dok/api/adresse#s%C3%B8gning.
+
+     *
+     * @param string $address
+     * @param string $path
+     *
+     * @throws AddressException
+     */
+    private function fetchAddressData(Address $address): array
+    {
+        try {
+            $client = HttpClient::create([
+                'base_uri' => 'https://api.dataforsyningen.dk/adresser',
+            ]);
+
+            $response = $client->request('GET', '', [
+                'query' => [
+                    'postnr' => $address->getPostalCode(),
+                    'vejnavn' => $address->getStreet(),
+                    'husnr' => $address->getNumber(),
+                    'etage' => $address->getFloor(),
+                    'dør' => $address->getSide(),
+                ],
+            ]);
+
+            $data = $response->toArray();
+
+            if (1 === count($data)) {
+                return reset($data);
+            }
+
+            if (count($data) > 1) {
+                throw $this->createException(sprintf('Ambiguous address: %s', (string) $address));
+            }
+        } catch (\Throwable $throwable) {
+            throw $this->createException(sprintf('Invalid address', (string) $address), $throwable->getCode(), $throwable);
+        }
+
+        throw $this->createException(sprintf('Invalid address: %s', (string) $address));
+    }
+
+    public function preUpdate(LifecycleEventArgs $args)
+    {
+        $object = $args->getObject();
+        // Set validatedAt to null when changing embedded Address entities.
+        $changeSet = $args->getEntityChangeSet();
+        foreach ($changeSet as $propertyPath => $value) {
+            $path = explode('.', $propertyPath, 2);
+            if ($this->propertyAccessor->isReadable($object, $path[0])) {
+                $embedded = $this->propertyAccessor->getValue($object, $path[0]);
+                // Only set validatedAt if not already being set in the change set.
+                if ($embedded instanceof Address && 'validatedAt' !== ($path[1] ?? null)) {
+                    $embedded->setValidatedAt(null);
+                }
+            }
+        }
+    }
```

Comment:
Would all this fit better into a separate `CaseListener`? If im not mistaken this will be executed even if we update an Agenda?

Reply:
You're right that this will run for all entity updates. Doing it like this (rather than in a CaseListener) makes it work for *all* embedded Address entities, i.e. if we ever add addresses outside case entities this will handle these addresses as well.